accepted/new-api.md

```diff
@@ -0,0 +1,971 @@
+    fn exports(&self, cx: impl AsContext) -> impl ExactSizeIterator<Item = Export>;
+    fn get_export(&self, cx: impl AsContext, name: &str) -> Option<Extern>;
+    fn get_func(&self, cx: impl AsContext, name: &str) -> Option<Func>;
+    fn get_typed_func<P, R>(&self, cx: impl AsContext, name: &str) -> Result<TypedFunc<P, R>>;
+    fn get_table(&self, cx: impl AsContext, name: &str) -> Option<Table>;
+    fn get_memory(&self, cx: impl AsContext, name: &str) -> Option<Memory>;
+    fn get_global(&self, cx: impl AsContext, name: &str) -> Option<Global>;
+}
+```
+
+### `Caller`
+
+```rust
+impl<'a, T> Caller<'a, T> {
+    fn context(&self) -> StoreContext<'_, T>;
+    fn context_mut(&mut self) -> StoreContextMut<'_, T>;
+}
+
+impl<T> AsContext for Caller<'a, T> {}
+impl<T> AsContextMut for Caller<'a, T> {}
+```
+
+### `Linker`
+
+```rust
+impl<'a, T> Linker<'a, T> {
+    fn new(cx: StoreContextMut<'a, T>) -> Linker<'a, T>;
+    fn new_with_host(store: StoreContextMut<'a, T>, host: &'a HostFunctions<T>) -> Linker<'a, T>;
+    fn define(&mut self, module: &str, name: &str, id: impl Into<Extern>) -> Result<()>;
+    fn define_name(&mut self, name: &str, id: impl Into<Extern>) -> Result<()>;
+    fn func(&mut self, module: &str, name: &str, func: impl IntoFunc) -> Result<()>;
+    fn instance(&mut self, name: &str, instance: &Instance) -> Result<()>;
+    fn module(&mut self, name: &str, module: &Module) -> Result<()>;
+    fn alias(/*...*/);
+    fn instantiate(&mut self, module: &Module) -> Result<Instance>;
+    async fn instantiate_async(&mut self, module: &Module) -> Result<Instance>;
+    fn iter(&self) -> impl Iterator<Item = (&str, &str, Extern)>;
+    fn get(&self, ty: &ImportType<'_>) -> Option<Extern>;
+    fn get_by_name(&self, module: &str, name: Option<&str>) -> impl Iterator<Item = Extern>;
+    fn get_one_by_name(&self, module: &str, name: Option<&str>) -> Result<Extern>;
+    fn get_default(&self, module: &str) -> Result<Func>;
+}
+
+impl<T> AsContext for Linker<'_, T> {}
+impl<T> AsContextMut for Linker<'_, T> {}
+```
+
+### `HostFunctions`
+
+```rust
+impl<T> HostFunctions<T> {
+    fn builder(engine: &Engine) -> HostFunctionsBuilder<T>;
+}
+
+impl<T> HostFunctionsBuilder<T> {
+    fn define(
+        &mut self,
+        module: &str,
+        name: &str,
+        ty: FuncType,
+        func: impl Fn(Caller<'_, T>, &[Val], &mut [Val]) -> Result<(), Trap> + Send + Sync + 'static,
+    );
+    fn wrap(&mut self, module: &str, name: &str, func: impl IntoFunc);
+
+    // + async versions ...
+
+    fn build(&mut self) -> HostFunctions<T>; // finalize & wrap in Arc
+}
+```
+
+### Examples
+
+To get an idea of what this API looks like in action, here's some examples
+translated from the main `wasmtime` repository's examples:
+
+#### "Hello, world!"
+
+```rust
+fn main() -> Result<()> {
+    let mut store: Store<()> = Store::default();
+    let module = Module::from_file(store.context().engine(), "examples/hello.wat")?;
```

Comment:
I'm honestly not entirely sure about what to do here. There's three objects in play: `Store`, `StoreContext`, and `StoreContextMut`. `Store` and `StoreContextMut` in theory can have the exact same APIs, and `StoreContext` can have everything that only requires shared access (like this method `engine()`).

Duplicating API surface area feels sort of bad to me though. Similarly though it feels bad to not be able to call `store.engine()`. It'd be nice if this could be done with `Deref` but that's not possible here since a `StoreContext` isn't itself borrowed from within the `Store`, it's a separate object.

I feel like there's downsides to both strategies basically. Duplicating the APIs is unfortunate and also not having them duplicated is unfortunate. I suspect though that duplication is probably the route to go since it's just a burden on us, the authors of `wasmtime`, and not consumers.

Reply:
I think I agree yeah, we should be optimizing for users here not us heh.